### PR TITLE
Create SingleRankPartitioner

### LINKED ISSUE #18729: Allow Master Apps to run on sub-communicators
## Reason
There are times when the master app should not be spread over very many MPI procs.  For instance, if running a very small/fast calculation as the master app and heavier apps as sub-apps.  Therefore it would be good to allow the master/main app to run on a small sub-communicator while still passing the full communicator to the MultiApp system to parcel out to sub-apps.

## Design
Add an `[App]` block to the input file that has a `max_procs` option in it (to mirror what is used in MultiApps).

## Impact
New capability.

### PR DESCRIPTION
**Pull request by friedmud**
Create a new partitioner that allows you to restrict all the elements to a single rank (no matter how many MPI you are using).

This is useful for restricting an app to run on only one processor.  For instance, you may want to restrict the master app - while having a large pool of processors to use for sub-apps.  This was requested by @aprilnovak .

closes #18729

### CODE REVIEW

File: framework/include/partitioner/SingleRankPartitioner.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+#pragma once
+
+// MOOSE includes
+#include "MoosePartitioner.h"
+
+class SingleRankPartitioner;
+
+template <>
+InputParameters validParams<SingleRankPartitioner>();
```
Comment (loganharbour) [inline framework/include/partitioner/SingleRankPartitioner.h:18]:
```suggestion
```

File: framework/include/partitioner/SingleRankPartitioner.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Partitions a mesh to one MPI rank
+ */
+class SingleRankPartitioner : public MoosePartitioner
+{
+public:
+  SingleRankPartitioner(const InputParameters & params);
+  virtual ~SingleRankPartitioner();
```
Comment (loganharbour) [inline framework/include/partitioner/SingleRankPartitioner.h:27]:
```suggestion
```

why?

File: framework/src/partitioner/SingleRankPartitioner.C [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "MooseApp.h"
+#include "MooseMesh.h"
+#include "MooseRandom.h"
+
+#include "libmesh/elem.h"
+
+registerMooseObject("MooseApp", SingleRankPartitioner);
+
+defineLegacyParams(SingleRankPartitioner);
```
Comment (loganharbour) [inline framework/src/partitioner/SingleRankPartitioner.C:20]:
```suggestion
```

File: framework/src/partitioner/SingleRankPartitioner.C [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+SingleRankPartitioner::SingleRankPartitioner(const InputParameters & params)
+  : MoosePartitioner(params), _rank(getParam<processor_id_type>("rank"))
+{
+  if (_rank >= _communicator.size())
+    paramError("rank", "Cannot be larger than the available number of MPI ranks");
+}
+
+SingleRankPartitioner::~SingleRankPartitioner() {}
+
```
Comment (loganharbour) [inline framework/src/partitioner/SingleRankPartitioner.C:42]:
```suggestion
```

File: framework/src/partitioner/SingleRankPartitioner.C [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+std::unique_ptr<Partitioner>
+SingleRankPartitioner::clone() const
+{
+  return libmesh_make_unique<SingleRankPartitioner>(_pars);
+}
+
+void
+SingleRankPartitioner::_do_partition(MeshBase & mesh, const unsigned int /*n*/)
+{
+  // Random number is on [0, 1]: scale to number of procs and round down
```
Comment (loganharbour) [inline framework/src/partitioner/SingleRankPartitioner.C:52]:
```suggestion
```

Thanks for copying from my random partitioner 😆

File: test/tests/partitioners/single_rank_partitioner/single_rank_partitioner.i [hunk 6]
```diff
@@ -0,0 +1,53 @@
+[Mesh]
```
Comment (loganharbour) [inline test/tests/partitioners/single_rank_partitioner/single_rank_partitioner.i:1]: Why do we always do simple diffusion on this?

I say just do a boring test with
```
[Problem]
  solve = false
[]

[Executioner]
  type = Steady
[]

File: test/tests/partitioners/single_rank_partitioner/tests [hunk 7]
```diff
@@ -0,0 +1,24 @@
+[Tests]
```
Comment (loganharbour) [inline test/tests/partitioners/single_rank_partitioner/tests:1]: Design and issues can be global:
```
[Tests]
  issues = '#18729'
  design = 'partitioners/SingleRankPartitioner.md'
  [test]
  ...
[]
```

File: test/tests/partitioners/single_rank_partitioner/tests [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  [./test]
+    type = 'Exodiff'
+    input = 'single_rank_partitioner.i'
+    exodiff = 'single_rank_partitioner_out.e'
+    min_parallel = 4
```
Comment (loganharbour) [inline test/tests/partitioners/single_rank_partitioner/tests:6]:
```suggestion
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/doc/content/source/partitioner/SingleRankPartitioner.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# SingleRankPartitioner
+
+!syntax description /Mesh/Partitioner/SingleRankPartitioner
+
+## Description
+
+Partitions all elements to the given `rank`.  Useful for restricting an application to run on only one processor.
```
Comment (loganharbour) [inline framework/doc/content/source/partitioner/SingleRankPartitioner.md:7]:
```suggestion
Partitions all elements to the given [!param](/Mesh/Partitioner/SingleRankPartitioner/rank).  Useful for restricting an application to run on only one processor.
```

File: framework/src/partitioner/SingleRankPartitioner.C [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SingleRankPartitioner.h"
+
+#include "MooseApp.h"
+#include "MooseMesh.h"
+#include "MooseRandom.h"
+
```
Comment (loganharbour) [inline framework/src/partitioner/SingleRankPartitioner.C:15]:
```suggestion
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: test/tests/partitioners/single_rank_partitioner/tests (hunk 8 above)
Comment (friedmud) [reply test/tests/partitioners/single_rank_partitioner/tests:6]: No - the min parallel is on purpose... the test puts the solve on rank 2

**Comment by GiudGiud** [comment]
If we're going to enable having more processes on the subapps than the main app, maybe we want a LimitedRankPartititioner rather than just single

**Comment by friedmud** [comment]
> If we're going to enable having more processes on the subapps than the main app, maybe we want a ReducedRankPartititioner rather than just single

"Reduced" is too hard right now.  Possible... but way more complicated than what's needed now.

Restriction to 1 is very common and will help a lot of people _right now_.

File: test/tests/partitioners/single_rank_partitioner/tests (hunk 8 above)
Comment (loganharbour) [reply test/tests/partitioners/single_rank_partitioner/tests:10]: Fine fine. but why not 2 then?

File: test/tests/partitioners/single_rank_partitioner/tests [hunk 11]
```diff
@@ -0,0 +1,20 @@
+[Tests]
+  issues = '#18729'
+  design = 'partitioners/SingleRankPartitioner.md'
+  requirement = 'The system shall error if an out of range rank is given to SingleRankPartitioner'
```
Comment (loganharbour) [inline test/tests/partitioners/single_rank_partitioner/tests:4]: Ah - this won't quite work though. You need to use the detail syntax; you cannot have duplicate requirements for a single test

Comment (friedmud) [reply test/tests/partitioners/single_rank_partitioner/tests:4]: Yeah - just me being hasty.  Let me fix it.

**Review by loganharbour — DISMISSED** [review]

**Comment by friedmud** [comment]
@loganharbour bah - it's because the `partitioner` directory is one of the dumb ones that isn't plural....

**Review by loganharbour — APPROVED** [review]

**Pull request merged**